once/_iterator_wrappers.py

```diff
@@ -0,0 +1,186 @@
+                await asyncio.sleep(0)
+                waiting_for_generating = False
+            async with self.lock:
+                if i == len(self.results) and not self.finished:
+                    if self.generating:
+                        # We're at the lead, but someone else is generating the next value
+                        # so we just hop back onto the next iteration of the loop
+                        # until it's ready.
+                        waiting_for_generating = True
+                        continue
+                    # We're at the lead and no one else is generating, so we need to increment
+                    # the iterator. We just store the value in self.results so that
+                    # we can later yield it outside of the lock.
+                    assert self.generator is not None
+                    # TODO(matt): Is the fact that we have to suppress typing here a bug?
+                    self.generating = self.generator.asend(send)  # type: ignore
+                    generating = self.generating
+                elif i == len(self.results) and self.finished:
+                    # All done.
+                    return
+                else:
+                    # We already have the correct result, so we grab it here to
+                    # yield it outside the lock.
+                    next_val = self.results[i]
+
+            if generating:
+                try:
+                    next_val = await generating
+                except StopAsyncIteration:
+                    async with self.lock:
+                        self.generator = None  # Allow this to be GCed.
+                        self.finished = True
+                        self.generating = None
+                        generating = None
+                        return
+                async with self.lock:
+                    self.results.append(next_val)
+                    generating = None
+                    self.generating = None
+
+            send = yield next_val
+            i += 1
+
+
+class GeneratorWrapper:
+    """Wrapper around an sync generator which only runs once.
+
+    Subsequent calls will return results from the first call, which is
+    evaluated lazily.
+    """
+
+    def __init__(self, func, *args, **kwargs) -> None:
+        self.generator: collections.abc.Generator | None = func(*args, **kwargs)
+        self.finished = False
+        self.results: list = []
+        self.generating = False
+        self.lock = threading.Lock()
+        self.next_send = None
+
+    def _yield_results(self) -> collections.abc.Generator:
+        i = 0
+        # Fast path for subsequent calls will not require a lock
+        while True:
+            # If we on before the penultimate entry, we can return now. When yielding the last
+            # element of results, we need to be recording next_send, so that needs the lock.
+            if i < len(self.results) - 1:
+                yield self.results[i]
```

Comment:
This is limiting us to cPython by assuming there is a GIL that makes this list thread-safe to access. Is that a limitation you are okay with? Let's put a comment either way about the assumption.